### PR TITLE
Fixed #3 (Make it build with ghcjs)

### LINKED ISSUE #3: Support building with GHCJS
cc @dcoutts @kosmikus 

Hey guys!

I'm trying (perhaps with a bit of foolish, Xmas spirit!) to install IRIS' big-kitchen-sink-restful-server using GHCJS, and as it depends upon `binary-serialise-cbor` I'm hitting [an issue similar to the hashable one](https://github.com/ghcjs/ghcjs-boot/blob/master/patches/hashable.patch) which has to do with unboxed constructors:

```
    Data/Binary/Serialise/CBOR/ByteOrder.hs:202:46:
        Couldn't match expected type ‘Word#’ with actual type ‘Word64#’
        In the first argument of ‘wordToFloat64#’, namely ‘w#’
        In the first argument of ‘D#’, namely ‘(wordToFloat64# w#)’

    Data/Binary/Serialise/CBOR/ByteOrder.hs:209:40:
        Couldn't match expected type ‘Word64#’ with actual type ‘Word#’
        In the third argument of ‘writeWord64Array#’, namely ‘w#’
        In the expression: writeWord64Array# mba# 0# w# s'
```

I suspect here the problem might lie in the different word size between the native vs the JS world. Do you guys think it's possible to issue a patch [with a sane dose of CPP](https://github.com/ghcjs/ghcjs-boot/blob/master/patches/hashable.patch#L10) to make the package buildable on ghcjs? :wink:  Is even possible at all?

Thanks a ton!

A.

### PR DESCRIPTION
**Pull request by adinapoli**
cc @kosmikus @dcoutts 

See #3 for context.

I have tested this both in ghcjs and with normal GHC compilations, and both works.

### CODE REVIEW

File: Data/Binary/Serialise/CBOR/Read.hs [hunk 1]
```diff
@@ -1177,7 +1177,11 @@ tryConsumeDouble :: Word8 -> ByteString -> DecodedToken Double
 tryConsumeDouble hdr !bs = case fromIntegral hdr :: Word of
   0xf9 -> DecodedToken 3 (float2Double $ wordToFloat16 $ withBsPtr grabWord16 (BS.unsafeTail bs))
   0xfa -> DecodedToken 5 (float2Double $ wordToFloat32 $ withBsPtr grabWord32 (BS.unsafeTail bs))
+#if !defined(ghcjs_HOST_OS)
   0xfb -> DecodedToken 9                (wordToFloat64 $ withBsPtr grabWord64 (BS.unsafeTail bs))
+#else
+  0xfb -> DecodedToken 9                (float2Double $ wordToFloat32 $ withBsPtr grabWord32 (BS.unsafeTail bs))
```
Comment (dcoutts) [inline Data/Binary/Serialise/CBOR/Read.hs]: Unfortunately this is not going to work. We have to read the full 8 bytes. Reading the first 4 will just give us nonsense.

We'll need to work out how to read an IEEE double from 8 bytes on GHCJS.

**Comment by dcoutts** [comment]
So actually I suspect we would hit the same compile failure on 32bit x86. We ought to add an impl of wordToFloat64 that works for Word64#, rather than Word# to use on 32bit arches + ghcjs.

**Comment by adinapoli** [comment]
Ah, that makes sense @dcoutts , thanks! I will try to investigate a bit and in case cook up something!

**Comment by adinapoli** [comment]
@dcoutts Sorry if what I'm pull requesting here is stupid, but I guess it's a cheeky attempt to see if we can get away with the "easy way".

I have amended the PR to read `Word`, changing the type appropriately (type tetris stuff, no clever implementation).

Could you tell me if you think this is equally wrong and, if it is, suggest a course of action on how we could tackle this?

Thanks a ton!

Alfredo

File: Data/Binary/Serialise/CBOR/ByteOrder.hs [hunk 2]
```diff
@@ -210,6 +211,21 @@ wordToFloat64# w# =
...
               (# _, f# #) -> f#
+#else
+{-# INLINE wordToFloat64 #-}
+wordToFloat64 :: Word -> Double
+wordToFloat64 (W# w#) = D# (wordToFloat64# w#)
+
+{-# NOINLINE wordToFloat64# #-}
+wordToFloat64# :: Word# -> Double#
+wordToFloat64# w# =
+    case newByteArray# 8# realWorld# of
```
Comment (adinapoli) [inline Data/Binary/Serialise/CBOR/ByteOrder.hs]: @dcoutts Ah, suspect: We are allocating 8 bytes here, but we need four. Do you think the above is simply a waste of space or something semantically wrong?

**Comment by dcoutts** [comment]
@adinapoli Can you check if the current HEAD version builds with ghcjs. I think there's a good chance that it will.

The current code there is:

    {-# INLINE wordToFloat64 #-}
    wordToFloat64 :: Word64 -> Double
    wordToFloat64 (W64# w#) = D# (wordToFloat64# w#)

    {-# NOINLINE wordToFloat64# #-}
    #ifdef ARCH_64bit
    wordToFloat64# :: Word# -> Double#
    #else
    wordToFloat64# :: Word64# -> Double#
    #endif
    wordToFloat64# w# =
        case newByteArray# 8# realWorld# of
          (# s', mba# #) ->
            case writeWord64Array# mba# 0# w# s' of
              s'' ->
                case readDoubleArray# mba# 0# s'' of
                  (# _, f# #) -> f#

Now presumably ghcjs classifies as 32bit, so it's in the second case above. On 32bit arches, there is a `Word64#` type (on 64bit arches there is just `Word#`), and the `W64#` constructor expects a `Word64#` (on 64bit arches it expects as `Word#`). So as far as I can tell, the above should just work, though I admit I've not yet tested it on x86 32bit.

**Comment by adinapoli** [comment]
Hey @dcoutts ,

I'm happy to report that this is indeed the case, thank you!

In order to improve on my own mistakes, would you spend 2 mins telling me how and why my last attempt (see https://github.com/adinapoli/binary-serialise-cbor/commit/c1476eeddbfa0d46e39882f7bafc066fb0be8353) was naive? That would be really appreciated!

Thanks!
A.

**Comment by dcoutts** [comment]
Ok, so here's that version:

    #if !defined(ghcjs_HOST_OS)
    {-# INLINE wordToFloat64 #-}
    wordToFloat64 :: Word64 -> Double
    wordToFloat64 (W64# w#) = D# (wordToFloat64# w#)

    {-# NOINLINE wordToFloat64# #-}
    wordToFloat64# :: Word# -> Double#
    wordToFloat64# w# =
        case newByteArray# 8# realWorld# of
          (# s', mba# #) ->
            case writeWord64Array# mba# 0# w# s' of
              s'' ->
                case readDoubleArray# mba# 0# s'' of
                  (# _, f# #) -> f#
    #else
    {-# INLINE wordToFloat64 #-}
    wordToFloat64 :: Word -> Double
    wordToFloat64 (W# w#) = D# (wordToFloat64# w#)

    {-# NOINLINE wordToFloat64# #-}
    wordToFloat64# :: Word# -> Double#
    wordToFloat64# w# =
        case newByteArray# 8# realWorld# of
          (# s', mba# #) ->
            case writeWordArray# mba# 0# w# s' of
              s'' ->
                case readDoubleArray# mba# 0# s'' of
                  (# _, f# #) -> f#
    #endif

and

    #if !defined(ghcjs_HOST_OS)
      0xfb -> DecodedToken 9                (wordToFloat64 $ withBsPtr grabWord64 (BS.unsafeTail bs))
    #else
      0xfb -> DecodedToken 9                (wordToFloat64 $ withBsPtr grabWord32 (BS.unsafeTail bs))
    #endif

So for starters, this second bit is wrong. Notice that you're grabbing a Word32 rather than a Word64. This cannot work. The binary format we are decoding really is a 64bit word, that we want to convert into a 64bit double float. We can't just grab the first 32bits of that 64bit value. That cannot decode correctly.

The other thing that's odd is that you're changing the type of `wordToFloat64`. This is odd because it's an exported function, so it ought to be presenting an arch-independent API, as all the other exported functions do.

Finally, the ghcjs specific version you've got is then wrong because it's only working with 32bits, rather than 64bits. It takes a `Word#` which is 32bits on 32bit arches (like ghcjs). Then it writes with `writeWordArray#` only 32bits, but then immediately reads 64bits with `readDoubleArray#`, so it will pick up 32bits of garbage. So it's not converting the double value correctly.

So, what things to think about here. This is low level code so you need to think about what the size of these types are. Are they 32bit or 64bit, and are they always 32 or 64 bit, or does that change depending on the arch? And in this context, ghcjs considers its target arch to be 32bit.

But also, what is the size of the thing we're decoding from binary? And notice that that size does not change. The external format is fixed.

The size of the `Word#` type varies with the arch. The `Word64#` type is a funny one. It only exists on 32bit arches. It's there to be the 64bit version for the 32bit arches where `Word#` is 32bit. So that means the `Word64` (no `#`) type contains a `Word#` on 64bit arches, but a `Word64#` on 32bit arches.

**Comment by adinapoli** [comment]
> So for starters, this second bit is wrong. Notice that you're grabbing a Word32 rather than a Word64. This cannot work.

Argh, you are perfectly right. I was **convinced** to have removed that stupid `grabWord32` from the 2nd commit and it turns out I did not. :cry: 

> This is odd because it's an exported function, so it ought to be presenting an arch-independent API, as all the other exported functions do.

Extremely well made point, I never thought of that from a library writer perspective. Great comment!

> The size of the Word# type varies with the arch. The Word64# type is a funny one. It only exists on 32bit arches. It's there to be the 64bit version for the 32bit arches where Word# is 32bit. So that means the Word64 (no #) type contains a Word# on 64bit arches, but a Word64# on 32bit arches.

> On the other hand, Double# is always 64bit, irrespective of arch.

Ah, it's definitely more fiddly than I thought originally!

Thank you @dcoutts , this feedback means the world to me!

A.

**Pull request closed without merging**